pkg/controller/job/backoff_utils.go

```diff
@@ -0,0 +1,167 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package job
+
+import (
+	"fmt"
+	"sort"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/utils/clock"
+)
+
+type backoffInfo struct {
+	key                      string
+	failuresAfterLastSuccess int32
+	lastFailureTime          *metav1.Time
+}
+
+type backoffInfoStore struct {
+	store cache.Store
+}
+
+func (s *backoffInfoStore) getBackoffInfo(key string) *backoffInfo {
```

Comment:
I suggest to inline this function in the place where it is used, as this is the only use for now.